// ejv/build/tester.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.regExpTester = exports.uniqueItemsTester = exports.arrayTypeOfTester = exports.arrayTester = exports.exclusiveMaxDateTester = exports.maxDateTester = exports.exclusiveMinDateTester = exports.minDateTester = exports.dateTester = exports.hasPropertyTester = exports.objectTester = exports.dateTimeFormatTester = exports.timeFormatTester = exports.dateFormatTester = exports.emailTester = exports.stringRegExpTester = exports.stringTester = exports.exclusiveMaxNumberTester = exports.maxNumberTester = exports.exclusiveMinNumberTester = exports.minNumberTester = exports.indexTester = exports.integerTester = exports.numberTester = exports.booleanTester = exports.maxLengthTester = exports.minLengthTester = exports.lengthTester = exports.enumTester = exports.definedTester = exports.typeTester = void 0;
var constants_1 = require("./constants");
var typeTester = function (value, type) {
    var valid;
    switch (type) {
        case constants_1.DataType.BOOLEAN:
            valid = exports.booleanTester(value);
            break;
        case constants_1.DataType.NUMBER:
            valid = exports.numberTester(value);
            break;
        case constants_1.DataType.STRING:
            valid = exports.stringTester(value);
            break;
        case constants_1.DataType.OBJECT:
            valid = exports.objectTester(value);
            break;
        case constants_1.DataType.DATE:
            valid = exports.dateTester(value);
            break;
        case constants_1.DataType.REGEXP:
            valid = exports.regExpTester(value);
            break;
        case constants_1.DataType.ARRAY:
            valid = exports.arrayTester(value);
            break;
    }
    return valid;
};
exports.typeTester = typeTester;
var definedTester = function (value) {
    return value !== undefined;
};
exports.definedTester = definedTester;
var enumTester = function (value, arr) {
    return arr.includes(value);
};
exports.enumTester = enumTester;
var lengthTester = function (value, length) {
    return value.length === length;
};
exports.lengthTester = lengthTester;
var minLengthTester = function (value, minLength) {
    return value.length >= minLength;
};
exports.minLengthTester = minLengthTester;
var maxLengthTester = function (value, maxLength) {
    return value.length <= maxLength;
};
exports.maxLengthTester = maxLengthTester;
var booleanTester = function (value) {
    return typeof value === 'boolean';
};
exports.booleanTester = booleanTester;
var numberTester = function (value) {
    return typeof value === 'number' && !isNaN(value);
};
exports.numberTester = numberTester;
var integerTester = function (value) {
    return +value.toFixed(0) === value;
};
exports.integerTester = integerTester;
var indexTester = function (value) {
    return exports.integerTester(value) && value >= 0;
};
exports.indexTester = indexTester;
var minNumberTester = function (value, min) {
    return value >= min;
};
exports.minNumberTester = minNumberTester;
var exclusiveMinNumberTester = function (value, min) {
    return value > min;
};
exports.exclusiveMinNumberTester = exclusiveMinNumberTester;
var maxNumberTester = function (value, max) {
    return value <= max;
};
exports.maxNumberTester = maxNumberTester;
var exclusiveMaxNumberTester = function (value, max) {
    return value < max;
};
exports.exclusiveMaxNumberTester = exclusiveMaxNumberTester;
var stringTester = function (value) {
    return typeof value === 'string';
};
exports.stringTester = stringTester;
var stringRegExpTester = function (value, regExp) {
    var valid = false;
    var _regExp = undefined;
    if (exports.regExpTester(regExp)) {
        _regExp = regExp;
    }
    else if (exports.stringTester(regExp)) {
        _regExp = new RegExp(regExp);
    }
    if (!!_regExp && exports.regExpTester(_regExp)) {
        valid = _regExp.test(value);
    }
    return valid;
};
exports.stringRegExpTester = stringRegExpTester;
// RFC 5322, 3.4.1. spec
var emailTester = function (value) {
    var valid = false;
    if (exports.stringTester(value) && exports.stringRegExpTester(value, /^.+@.+$/)) {
        var valueAsString = value;
        var atIndex = valueAsString.lastIndexOf('@');
        var localPart = valueAsString.substr(0, atIndex);
        var domain = valueAsString.substr(atIndex + 1);
        // regular expression sources
        // const aTextRegExpStr : string = '[-a-zA-Z0-9!#$%&\\\'*+/=?^_`{|}~]+';
        var dotAtomRegExp = /^(\.?[-a-zA-Z0-9!#$%&'*+/=?^_`{|}~]+)*$/;
        var quotedStringRegExp = /^"[\u0020-\u005b\u005d-\u007e\\]*"$/; // include space (\u005b)
        var domainLiteralRegExp = /^\[[\u0020-\u005a\u005c-\u007e\\]*]$/;
        var validLocalPart = localPart.length <= 64
            && (dotAtomRegExp.test(localPart)
                || quotedStringRegExp.test(localPart));
        var validDomain = !domain.startsWith('.') && !domain.endsWith('.')
            && (dotAtomRegExp.test(domain)
                || domainLiteralRegExp.test(domain));
        valid = validLocalPart && validDomain;
    }
    return valid;
};
exports.emailTester = emailTester;
// RFC 3339 (https://www.ietf.org/rfc/rfc3339.txt) : YYYY-MM-DDThh:mm:ss[.SSSZ]
var rfc3339Tester = function (value) {
    return /^[0-9]{4}-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])T([0-1][0-9]|2[0-3])(:([0-5][0-9])){2}(\.\d+)?(Z|[-+]\d{2}:\d{2})?$/.test(value);
};
var iso8601DateTester = function (value) {
    var years = '(\\d{4})';
    var months = '(0[1-9]|1[0-2])';
    var dates = '(0[1-9]|[1-2][0-9]|3[0-1])';
    var dateOfYear = '(00[1-9]|0[1-9][0-9]|[1-2]\\d{2}|3[0-5]\\d|36[0-6])'; // 366 for leap year
    var weeks = '(W(0[1-9]|[2-4][0-9]|5[0-3]))';
    var days = '[1-7]';
    return [
        new RegExp("^[-+]?" + years + "$"),
        new RegExp("^" + years + "-" + months + "(-" + dates + ")?$"),
        new RegExp("^" + years + months + dates + "$"),
        new RegExp("^--" + months + "-?" + dates + "$"),
        new RegExp("^" + years + "-" + weeks + "(-" + days + ")?$"),
        new RegExp("^" + years + weeks + "(" + days + ")?$"),
        new RegExp("^" + years + "-?" + dateOfYear + "$") // ordinal dates : YYYY-DDD, YYYYDDD
    ].some(function (regExp) {
        return regExp.test(value);
    });
};
var iso8601TimeTester = function (value) {
    var hours = '([0-1]\\d|2[0-3])';
    var minutes = '([0-5]\\d)';
    var seconds = '([0-5]\\d|60)'; // 60 for leap second
    var ms = '(\\.[0-9]+)';
    return [
        new RegExp("^(" + hours + "|24)$"),
        new RegExp("^((" + hours + ":" + minutes + ")|24:00)$"),
        new RegExp("^((" + hours + ":" + minutes + ":" + seconds + ")|24:00:00)$"),
        new RegExp("^((" + hours + ":" + minutes + ":" + seconds + ms + ")|24:00:00.0+)$"),
        new RegExp("^(" + hours + minutes + "|2400)$"),
        new RegExp("^(" + hours + minutes + seconds + "|240000)$"),
        new RegExp("^(" + hours + minutes + seconds + ms + "|240000.0+)$") // hhmmss.sss
    ]
        .some(function (regExp) {
        return regExp.test(value);
    });
};
var iso8601DateTimeTester = function (value) {
    var valid = false;
    if (/.+T.+/.test(value) // should have 1 'T'
        && /(Z|[-+]\d{2}:?\d{2})$/.test(value) // should end with 'Z' or timezone
    ) {
        var _a = value.split('T'), date = _a[0], time = _a[1];
        if (time.endsWith('Z')) {
            time = time.replace('Z', '');
        }
        else {
            var timezoneStartIndex = time.includes('+') ? time.indexOf('+') : time.indexOf('-');
            time = time.substr(0, timezoneStartIndex);
        }
        valid = iso8601DateTester(date) && iso8601TimeTester(time);
    }
    return valid;
};
var dateFormatTester = function (value) {
    return iso8601DateTester(value);
};
exports.dateFormatTester = dateFormatTester;
var timeFormatTester = function (value) {
    return iso8601TimeTester(value);
};
exports.timeFormatTester = timeFormatTester;
var dateTimeFormatTester = function (value) {
    return rfc3339Tester(value) || iso8601DateTimeTester(value);
};
exports.dateTimeFormatTester = dateTimeFormatTester;
// // with port
// export const urlTester = (value : any) : boolean => {
// 	return stringTester(value)
// 		&& /^(?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,}))\.?)(?::\d{2,5})?(?:[/?#]\S*)?$/i.test(value);
// };
//
// // TODO: with port
// export const ipv4Tester = (value : any) : boolean => {
// 	return stringTester(value)
// 		&& /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/.test(value);
// };
//
// export const ipv6Tester = (value : any) : boolean => {
// 	return stringTester(value)
// 		&& /^s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:)))(%.+)?s*(\/([0-9]|[1-9][0-9]|1[0-1][0-9]|12[0-8]))?$;/.test(value);
// };
//
// export const ipTester = (value : any) : boolean => {
// 	return ipv4Tester(value) || ipv6Tester(value);
// };
var objectTester = function (value) {
    return typeof value === 'object';
};
exports.objectTester = objectTester;
var hasPropertyTester = function (value) {
    return Object.keys(value).length > 0;
};
exports.hasPropertyTester = hasPropertyTester;
var dateTester = function (value) {
    return exports.objectTester(value)
        && value !== null
        && typeof value.getFullYear === 'function'
        && !isNaN(value.getFullYear());
};
exports.dateTester = dateTester;
var minDateTester = function (value, min) {
    return +value >= +min;
};
exports.minDateTester = minDateTester;
var exclusiveMinDateTester = function (value, min) {
    return +value > +min;
};
exports.exclusiveMinDateTester = exclusiveMinDateTester;
var maxDateTester = function (value, max) {
    return +value <= +max;
};
exports.maxDateTester = maxDateTester;
var exclusiveMaxDateTester = function (value, max) {
    return +value < +max;
};
exports.exclusiveMaxDateTester = exclusiveMaxDateTester;
var arrayTester = function (value) {
    return Array.isArray(value);
};
exports.arrayTester = arrayTester;
var arrayTypeOfTester = function (array, type) {
    return array.every(function (item) {
        return exports.typeTester(item, type);
    });
};
exports.arrayTypeOfTester = arrayTypeOfTester;
var uniqueItemsTester = function (array) {
    return array.every(function (item) {
        return array.filter(function (target) { return target === item; }).length === 1;
    });
};
exports.uniqueItemsTester = uniqueItemsTester;
var regExpTester = function (value) {
    return value instanceof RegExp;
};
exports.regExpTester = regExpTester;
